### ISSUE TITLE
Question on "in batch softmax" and retrieval task

### ISSUE THREAD

**Issue by Benjamin-OPC** [tensorflow/recommenders#279]
Hi,

Thanks all for this awesome library ! I have a question regarding the retrieval task and the way it's implemented. From what I understand from the source code, the retrieval task uses CategoricalCrossEntropy under the hood with a queries x candidates embeddings matrix multiplication to get the scores. All of these are compared with an identity matrix which acts as the y_true variable in the loss function.

Let's assume that we have 4 users and 4 products, and that the batch size is 2. Let's now assume our dataset looks like :

| User         | Product   | 
| ----------|:---------:|
|  A              | 1              | 
|  B              | 2              | 
|  A              | 2              | 
|  B              | 1              | 

For the first batch, the loss will try to optimize the following matrix:
[[1, 0],[0, 1]]

For the second batch, the loss will try to optimize the following matrix:
[[0, 1],[1, 0]]

which is the exact opposite of what the network was told at batch 1.

My question is therefore : how can we avoid "erasing" or "contradicting" interactions for users who come back in different batches ?

Thanks a lot for your help !

**Comment by maciejkula** [comment]
This is in general quite difficult - you could have billions of interactions you're training on, and keeping track of who interacted with what across the entire dataset is not really practical.

In practice, this tends not to matter much beyond a mild regularizing effect on the model.
